### PR TITLE
Loki query editor: update CompletionDataProvider time range when it changes

### PR DESCRIPTION
**Pull request by matyax**
Currently we were creating a CompletionDataProvider instance when the editor was mounted, with whatever timerange was used at that moment. With these changes, we update the `CompletionDataProvider` time range reference when the prop is updated, which unfortunately happens very often, so this solution was developed in such a way to minimize the amount of `/labels` calls that are made when we `.start()` the language provider.

Starting the LanguageProvider with the updated time range can be deferred to the moment where we need the updated labels: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts#L46

Explore also starts it when the time range changes, so we can rely on both to prevent any extra `/labels` request, which can be very expensive.

I double checked and `datasource.languageProvider` doesn't change unless you change the datasource, because I was worried about leaky detached instances.

Lastly, the `startedTimeRange` vs new `range` in `LanguageProvider` was updated because it was returning false negatives.

### CODE REVIEW

File: public/app/plugins/datasource/loki/LanguageProvider.test.ts [hunk 1]
```diff
@@ -57,11 +57,14 @@ describe('Language completion provider', () => {
...
     it('should again fetch labels on second start with different timerange', async () => {
       const languageProvider = new LanguageProvider(datasource);
+      jest.mocked(datasource.getTimeRangeParams).mockRestore();
       const fetchSpy = jest.spyOn(languageProvider, 'fetchLabels').mockResolvedValue([]);
       await languageProvider.start();
       expect(fetchSpy).toHaveBeenCalledTimes(1);
       await languageProvider.start(mockTimeRange);
       expect(fetchSpy).toHaveBeenCalledTimes(2);
+      await languageProvider.start(mockTimeRange);
+      expect(fetchSpy).toHaveBeenCalledTimes(2);
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/loki/LanguageProvider.test.ts:67]: I'm a bit confused as to what this is accomplishing, won't this test pass without any of the implementation in this PR?

Comment (matyax) [reply public/app/plugins/datasource/loki/LanguageProvider.test.ts:67]: Good question. I pass the same time range again, and the expectation is that it sees that it was already started at that time and it doesn't get called 3 times.

Comment (gtk-grafana) [reply public/app/plugins/datasource/loki/LanguageProvider.test.ts:67]: gotcha, so it's good to keep this in here, but we should probably add a unit test that shows that changing the time range and triggering another query will update the time range sent?

Comment (matyax) [reply public/app/plugins/datasource/loki/LanguageProvider.test.ts:67]: It's not a bad idea, but it actually uses the time range that you pass as an argument, that you can find in all the functions: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/LanguageProvider.ts#L253 , or the default time range, which even the `start()` method uses.

This whole `start()` thing is just to keep track of "when" it was started so it doesn't "restart".

The actual regression test would be to change the time range in `MonacoQueryField` and assert that `LanguageProvider` was updated, but we don't unit test the editor.

**Review by gtk-grafana — APPROVED** [review]
Works as expected locally, can verify the bug is fixed. LGTM

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts [hunk 2]
```diff
@@ -82,7 +87,9 @@ export class CompletionDataProvider {
         // Make room in the cache for the fresh result by deleting the "first" index
         const keys = this.queryToLabelKeysCache.keys();
         const firstKey = keys.next().value;
-        this.queryToLabelKeysCache.delete(firstKey);
+        if (firstKey) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:90]: Ahhhh, this is why it was failing with the empty query!

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:90]: Changed to !== undefined.

**Pull request merged**